core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java

```diff
@@ -105,26 +132,33 @@ public double getPbFrac() {
         return ((double) getBytes()) / ByteSizeUnit.C5;
     }
 
+    public String getStringRep() {
+        if (size <= 0) {
+            return String.valueOf(size);
+        }
+        return getBytes() + ByteSizeUnit.BYTES.getSuffix();
```

Comment:
Why not `size` and `unit`?

Reply:
We can do this but to maintain backward compatibility I think we need to keep the parse method converting the unit to bytes for 6.x and can only make it keep the unit in a followup PR for only the master branch. It also means that we can't compare the string representations in `BizeSizeValueTests#testParse`[1] since the parsed version will always be in bytes but the original will be in any unit. I think this is ok because we chack the original and the parsed version are equal but I wanted to point it out. Does this sound ok to you?


[1] https://github.com/elastic/elasticsearch/pull/27702/files#diff-6be961ed1a7543ed4fff562bef6b5680R240

Reply:
Actually I think I have got a bwc to change the parsing to preserve the unit as well. To keep it bwc I've made it revert to converting the unit to bytes if the input value is fractional.